File: app/lib/api/centrifugo.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../logger.dart';
import 'client.dart';

class CentrifugoTokenResponse {
  final String connectionToken;
  final String subscriptionToken;
  final String channel;
  CentrifugoTokenResponse({
    required this.connectionToken,
    required this.subscriptionToken,
    required this.channel,
  });
  factory CentrifugoTokenResponse.fromJson(Map<String, dynamic> j) =>
      CentrifugoTokenResponse(
        connectionToken: j['connectionToken'] as String,
        subscriptionToken: j['subscriptionToken'] as String,
        channel: j['channel'] as String,
      );
}

Future<CentrifugoTokenResponse> getCentrifugoToken() async {
  logApi.info('getCentrifugoToken');
  return withAuthRetry(() async {
    final r = await http.get(
      Uri.parse('$apiBaseUrl/api/centrifugo/token'),
      headers: apiHeaders,
    );
    checkAuthResponse(r, fallback: '获取连接凭证失败');
    final res = CentrifugoTokenResponse.fromJson(
      jsonDecode(r.body) as Map<String, dynamic>,
    );
    logApi.info('getCentrifugoToken success channel=${res.channel}');
    return res;
  });
}

File: web/src/lib/api/centrifugo.ts
import { logger } from '../logger';
import { getApiUrl, TAG, AuthError, getToken, isAuthFailure, withAuthRetry } from './client';

export type CentrifugoTokenResponse = {
  connectionToken: string;
  subscriptionToken: string;
  channel: string;
};

export async function getCentrifugoToken(): Promise<CentrifugoTokenResponse> {
  logger.info(TAG, 'getCentrifugoToken');
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const res = await fetch(`${getApiUrl()}/api/centrifugo/token`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'getCentrifugoToken failed', res.status);
      throw new Error('Failed to get Centrifugo token');
    }
    const data = await res.json() as CentrifugoTokenResponse;
    logger.info(TAG, 'getCentrifugoToken success channel=', data.channel);
    return data;
  });
}
